llama-index-core/llama_index/core/agent/react/step.py

```diff
@@ -87,13 +87,26 @@ def __init__(
         callback_manager: Optional[CallbackManager] = None,
         verbose: bool = False,
         tool_retriever: Optional[ObjectRetriever[BaseTool]] = None,
+        complaint_when_no_reasoning_step: str = "",
     ) -> None:
         self._llm = llm
         self.callback_manager = callback_manager or llm.callback_manager
         self._max_iterations = max_iterations
         self._react_chat_formatter = react_chat_formatter or ReActChatFormatter()
         self._output_parser = output_parser or ReActOutputParser()
         self._verbose = verbose
+        if complaint_when_no_reasoning_step:
+            # TODO(feature): Instead of always giving a static message, let developer extend
+            #  `_handle_failure_in_extract_reasoning_step` such that the message can point out exactly what didn't match
+            #  the expectation of `_extract_reasoning_step` / violated the System Prompt.
+            self.dummy_tool_output_when_no_reasoning_step = ToolOutput(
```

Comment:
tbh, I don't think a static message is quite as helpful

I think it would work well to pass in `handle_failure_in_extract_reasoning_step` as a function argument. And if not provided, it runs the current one as a fallback?

```
def handle_failure(exp):
  ...

agent = ReActAgent.from_tools(..., handle_reasoning_failure_fn=handle_failure)
```

Reply:
Updated. How does it look this time?

Note: I changed the default behavior from raising that error to handling it with that function. I hope this is OK.

Reply:
Looks good to me, thanks!